cpp/ql/src/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll

```diff
@@ -460,6 +460,39 @@ private predicate isRecursiveDef(RangeSsaDefinition def, StackVariable v) {
   defDependsOnDefTransitively(def, v, def, v)
 }
 
+/**
+ * Holds if the bounds of `e` depend on a recursive definition, meaning that
+ * `e` is likely to have many candidate bounds during the main recursion.
+ */
+private predicate isRecursiveExpr(Expr e) {
+  exists(RangeSsaDefinition def, StackVariable v | exprDependsOnDef(e, def, v) |
+    isRecursiveDef(def, v)
+  )
+}
+
+/**
+ * Holds if `binop` is a binary operation that's likely to be assigned a
+ * quadratic (or more) number of candidate bounds during the analysis. This can
+ * happen when two conditions are satisfied:
+ * 1. It is likely there are many more candidate bounds for `binop` than for
+ *    its operands. For example, the number of candidate bounds for `x + y`,
+ *    denoted here nbounds(`x + y`), will be O(nbounds(`x`) * nbounds(`y`)).
+ *    In contrast, nbounds(`b ? x : y`) is only O(nbounds(`x`) + nbounds(`y`)).
+ * 2. Both operands of `binop` are recursively determined and are therefore
+ *    likely to have a large number of candidate bounds.
```

Comment:
This seems like a very reasonable approach.

I'd like to know a little more about how widely performance is improved, and by how much?

I'd also like to find at least one case where accuracy is affected, to get an idea of what kind of effect it might have.

Reply:
As for performance, it looks like this change leads to an overall degradation in performance on the projects in CPP-Differences. Most are unchanged, but torvalds/linux and php/php-src get slightly slower. I tried re-running torvalds/linux locally, and I do see a 12s slowdown, 10 seconds of which can be explained by more time spent in the predicates I'm changing in this PR.

Looking at the tuple-count logs, I don't see much room for improvement. There is more work done now that there's a `min`/`max` aggregate in the hot loop, but AFAICT the aggregation is only applied to the tuples where its results will ultimately be used. It's possible that the deep if-then-else structures could be flattened into a more static partitioning of cases, but I think that would come with a readability tax we don't want to pay, splitting each of `getTruncated{Lower,Upper}Bounds` into at three or more predicates.

Reply:
I've added a test now. I rebased the PR so the first commit is the test and the second commit shows the test changes.

Notice that the range in the test example is already imprecise before the change to add more widening. This PR only adds widening to degrade the precision of binary operators whose operands were already imprecise because they were widened.

Reply:
Thanks, the example really clarified what's going on for me.